doc/decisions/0_drafts/transformations.md

```diff
@@ -0,0 +1,128 @@
+For Elektra, the key `/DISPLAY/BRIGHTNESS` does not exist, as the `rename` plugin transformed this into the lowercase `/display/brightness`.
+
+This leads to problems with the notification plugins.
+As notification plugins are executed after the post-storage phase of the `set` operation, they will receive a keyset with the already transformed keys.
+In this example, the notification plugins will receive all-UPPERCASE keys, and send out notifications with those all-UPPERCASE keys. 
+An application listening to those notifications will not be able to query Elektra for those keys, as for Elektra those UPPERCASE keys do not exist.
+
+Apart from the problems with notifications, the way key name changing plugins work also breaks change tracking in plugins like `dbus`.
+This is because key names are read-only when they are contained in a `KeySet`.
+In order to change the name of a key, such a plugin has to create a new key with the changed name, and delete the key with the old name.
+The `dbus` plugin implements change tracking by checking the 'key needs sync' flag instead of comparing the values.
+As new keys by design have the 'key needs sync' flag set, the plugins that implement change tracking via the flag will always erroneously detect transformed keys as changed.
+
+### Observed problems with changing values (normalization)
+
+Suppose we have a plugin that changes the value of a key to the hard-coded value `1` during the storage phase of the `set` operation.
+If a plugin does change tracking, this will lead to false positives.
+
+```
+user:/limits/openfiles = 1
+```
+
+If the user changes the value, e.g. using `kdb set user:/limits/openfiles 23`, plugins will observe the new value `23`.
+However, the value-changing plugin will reset that value back to `1`.
+So in practice, the configuration has not been changed.
+Plugins relying on change tracking plugins (e.g. notification plugins) will however think that it has.
+
+### Observed problems with adding and removing metadata
+
+This is a bit of a mixture between changing key names and changing values.
+
+## Constraints
+
+1.
+2.
+3.
+
+## Assumptions
+
+1. We want to enable some kind of transformations
+2.
+3.
+
+## Considered Alternatives
+
+### Enable setting and unsetting of the `keyNeedsSync` flag
+
+Require plugins that rename keys to remove the `keyNeedsSync` flag.
+This would require making the 'key sync' APIs public.
+This does not solve the problem of changing values.
+
+### Use change tracking to replace the `keyNeedsSync` flag
+
+Eliminate the need of the `keyNeedsSync` flag by utilizing the planned change tracking API.
+This does not solve the problem of changing values and false positives.
+
+### Store original names and values in metadata.
+
+For key name transformations require that transformation plugins set a metakey (e.g. `meta:/elektra/runtimename`) with the runtime name before they do any transformations in the `kdbSet` phase.
```

Comment:
I'd prefer, if we say plugins must use a special function to transform key names, e.g.:

```c
typedef const char * (*ElektraNameTransform)(const Key *);

int elektraApplyNameTransform (KeySet * ks, ElektraNameTransform transform);
```

This `elektraKeyTransformName` could set the `meta:/` key or mark the key in some other way. That makes things easier, if we discover the `meta:/` isn't enough and also avoids potential mistakes (like changing an existing `meta:/` key). The function could also use "hacky tricks" to avoid creating new keys, if the `KeySet * ks` holds the only reference.

We could have a similar function for the value of a key.

Reply:
Good idea! I added it.